server/channel_state.go

```diff
@@ -281,32 +275,35 @@ func (p *Plugin) cleanUpState() error {
 				continue
 			}
 
-			if err := p.cleanCallState(k); err != nil {
+			state, err := p.lockCall(k)
+			if err != nil {
+				p.LogError("failed to lock call", "err", err.Error())
+				continue
+			}
+			if err := p.cleanCallState(k, state); err != nil {
+				p.unlockCall(k)
 				return fmt.Errorf("failed to clean up state: %w", err)
 			}
+			p.unlockCall(k)
 		}
 		page++
 	}
 	return nil
 }
 
-func (p *Plugin) cleanCallState(channelID string) error {
-	if err := p.kvSetAtomicChannelState(channelID, func(state *channelState) (*channelState, error) {
-		if state == nil {
-			return nil, nil
-		}
-		state.NodeID = ""
+func (p *Plugin) cleanCallState(channelID string, state *channelState) error {
```

Comment:
Yeah we can certainly add a comment but at the end of the day some care must be taken when dealing with locks, there's no easy way around it. The only "protective" measure I can think of would be to implement a `TryLock()` method on the cluster mutex and send a big error if we get there and we are not locked. It's not pretty but doable if you think it's worth it.

Reply:
I think having comments is the least we can do, unless you can think of a way to isolate the need for locks? I'm not sure I'd want to change to a lock-passing design, but at least that would make the dependency explicit. It's tricky, for sure, that's why I'm trying to think of ways to make it less error prone.

Reply:
I am genuinely curious, could you elaborate a bit on the "lock-passing" design? What would `cleanCallState` do with the lock in this instance?

Reply:
In the case of `cleanCallState` it wouldn't need any change, because it doesn't release and reset the lock. And maybe that's part of the problem -- it doesn't look like it needs a lock, but it does. Maybe there's a better way to force that? Do you have any ideas?
If we went to lock-passing style, I would assume that the other cases where a function _does_ need one, it would take the lock object as a parameter to make the dependency explicit. 
Again, I'm not really sure if we should pass around a lock, I'm just trying to figure out a way to make it less likely that we'll use `cleanCallState` without holding one.

Reply:
@cpoile To be honest I can't see a huge problem here. It's not like calling `cleanCallState` without a lock would cause any critical issue per se, it's just not atomic and should be avoided. Any function that we call in a critical section would be affected by the same reasoning. Some times it may be perfectly fine to call the same method with or without lock.

That said, the only thing that I can think of to force what you suggest would be to inline the function. It's not the most complex function so it could be done but the tradeoff is repeating its logic in different places.

Reply:
@cpoile Let me know on the above.

Reply:
Ok, sounds good to me! :)